Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* #68522
* __->__ #69909
* #69908

This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

Differential Revision: [D33774422](https://our.internmc.facebook.com/intern/diff/D33774422)

lezcano committed:
Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True

This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

albanD commented on test/test_ops.py:
```diff
@@ -687,6 +687,29 @@ def test_composite_compliance(self, device, dtype, op):
 
+    @onlyCPU
+    @ops(op_db, allowed_dtypes=(torch.float,))
+    def test_floating_inputs_are_differentiable(self, device, dtype, op):
```
If this is just enforced and there is not way to make some inputs not require gradients, why don't we just set requires grad to True whatever the user returned?

Lezcano replied:
We want to be able to skip/xfail this for some operations, when the support is limited.

albanD replied:
Don't want a way for the user to specify that things explicitly shouldn't require gradients?
We could have a `mark_non_differentiable(some_tensor)` that users can use during the input creation.

And in that method we could set `some_tensor._not_requires_grad_on_purposes` and have this new check look for this field on the given Tensor?

Lezcano replied:
@mruberry preferred this test over marking individual tensors.

Lezcano commented on tools/autograd/derivatives.yaml:
```diff
@@ -733,6 +733,7 @@
 
 - name: index.Tensor(Tensor self, Tensor?[] indices) -> Tensor
   self: index_backward(grad.new_zeros(self.sizes(), self.options()), indices, grad)
+  result: self_t.index(indices)
```
@albanD Just to be in the clear, mind having a look at this one. I was getting an error because this one was not implemented so I implemented it.

On a different note, how cool would it be to have automatically implemented the forward derivatives for linear / multilinear functions...

albanD replied:
You can use `auto_linear` here?
And we could expand it to multilinear functions as well. multi-affine actually would be the best as most ops are affine in our case.
something like `auto_affine(linear=(linear_arg1, linear_arg2), affine=(bias_arg1, bias_arg2))`? If we find a nice way to express these, we can definitely add it.

Lezcano replied:
I think I got some error when using `auto_linear`.

And yes, extending it to `auto_affine` that'd be great. Now, I'd call the second term `constant` rather than  `affine`, but yep. I think that notation could certainly be general enough as to represent a good amount of PyTorch functions.

Lezcano replied:
Nah, you were right. Corrected.

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774422).

mruberry commented:
Internal tools are failing to merge the stack up to here, unfortunately, and are requesting a rebase.

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

Differential Revision: [D33774422](https://our.internmc.facebook.com/intern/diff/D33774422)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774422).

lezcano committed:
Update on "Add OpInfo test to check that floating point inputs in OpInfos have requires_grad set to True"


This test detected a number of sampling methods that were not generating
the samples as expected, e.g. `index_put`, `cosine_embedding`, `stft`, but
perhaps most notably the generator for `BinOps`.

It also detected that `reminder` and `fmod` did not have implemented the
backward formula for the second input. I added this in the previous PR.

Differential Revision: [D33774422](https://our.internmc.facebook.com/intern/diff/D33774422)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774422).
